jerry-core/ecma/base/ecma-globals.h

```diff
@@ -204,20 +204,99 @@ typedef enum
-  ECMA_PROPERTY_FLAG_LCACHED = 1u << 6, /**< property is lcached */
+  ECMA_PROPERTY_TYPE_DELETED, /**< deleted property */
+  ECMA_PROPERTY_TYPE_INTERNAL, /**< internal property */
+  ECMA_PROPERTY_TYPE_NAMEDDATA, /**< property is named data */
+  ECMA_PROPERTY_TYPE_NAMEDACCESSOR, /**< property is named accessor */
+
+  ECMA_PROPERTY_TYPE_HASH, /**< property hash (TODO: implement this) */
+
+  ECMA_PROPERTY_TYPE_PROPERTY_PAIR__MAX = ECMA_PROPERTY_TYPE_NAMEDACCESSOR, /**< highest value for
+                                                                             *   property pair types. */
+  ECMA_PROPERTY_TYPE__MAX = ECMA_PROPERTY_TYPE_HASH, /**< highest value for property types. */
+} ecma_property_types_t;
+
+/**
+ * Property type mask.
+ */
+#define ECMA_PROPERTY_TYPE_MASK 0x7
+
+/**
+ * Property flags base shift.
+ */
+#define ECMA_PROPERTY_FLAG_SHIFT 3
+
+/**
+ * Property flag list (for ECMA_PROPERTY_TYPE_NAMEDDATA
+ * and ECMA_PROPERTY_TYPE_NAMEDACCESSOR).
+ */
+typedef enum
+{
+  ECMA_PROPERTY_FLAG_CONFIGURABLE = 1u << (ECMA_PROPERTY_FLAG_SHIFT + 0), /**< property is configurable */
+  ECMA_PROPERTY_FLAG_ENUMERABLE = 1u << (ECMA_PROPERTY_FLAG_SHIFT + 1), /**< property is enumerable */
+  ECMA_PROPERTY_FLAG_WRITABLE = 1u << (ECMA_PROPERTY_FLAG_SHIFT + 2), /**< property is writable */
+  ECMA_PROPERTY_FLAG_LCACHED = 1u << (ECMA_PROPERTY_FLAG_SHIFT + 3), /**< property is lcached */
 } ecma_property_flags_t;
 
 /**
+ * Abstract property representation.
+ *
+ * A property is a type_and_flags byte and an ecma_value_t value pair.
+ * This pair is represented by a single pointer in JerryScript. Although
+ * a packed struct would only consume sizeof(ecma_value_t)+1 memory
+ * bytes, accessing such structure is inefficient from the CPU viewpoint
+ * because the value is not naturally aligned. To improve performance,
+ * type bytes and values are packed together. It ecma_value_t is four
+ * bytes long, maximum four properties can be packed together:
+ *
+ *  [Type 1][Type 2][Type 3][Type 4][Value 1][Value 2][Value 3][Value 4]
+ *
+ * This way no memory is wasted and values are naturally aligned.
+ *
+ * For property pairs, we only use two values:
+ *
+ *  [Type 1][Type 2][Unused 1][Unused 2][Value 1][Value 2]
+ *
+ * The unused bytes can be used for other purposes of course.
+ *
+ * The most important feature of this layout: the address of a value
+ * can be computed from the address of the type_and_flags byte.
+ *
+ * Property pointers cannot be compressed.
+ */
+typedef struct
+{
+  uint8_t type_and_flags; /**< ecma_property_types_t (3 bit) and ecma_property_flags_t */
+} ecma_property_t;
+
+/**
+ * Property header for all items in a property list.
+ */
+typedef struct
+{
+  ecma_property_t first_type; /**< first property type byte */
+  ecma_property_t second_type; /**< second property type byte
+                                *   might be unused for certain property types */
```

Comment:
I'n not sure about your reply here. I was asking why the comment of `second_type` says that it might be unused for some property types when this is not visible from the other parts of the PR?